src/iter/checked_product.rs

```diff
@@ -0,0 +1,95 @@
+use crate::identities::One;
+use crate::CheckedMul;
+
+/// This trait represents types of which an iterator can be multiplied up with overflow checking.
+/// This trait should rarely be called directly.
+pub trait CheckedProduct<Result = Self> {
+    /// Multiplies up the elements of an iterator, returning `None` if an overflow would occur.
+    ///
+    /// Multiplies up an empty iterator returns a value representing One.
+    ///
+    /// If the iterator contains Zero, the order of elements may effect whether the result is `None`.
+    fn checked_product<I: Iterator<Item = Self>>(iter: I) -> Option<Result>;
```

Comment:
This is opposite of how `Product` is defined, where the type parameter is the _input_ (`Item`) and it always outputs `Self`. I would rather match that for consistency, and same for `Sum`.

Reply:
I completely agree. I feel very silly for having done it the other way around. I have changed it now.